lib/client.rb

```diff
@@ -0,0 +1,81 @@
+class Client
+  attr_accessor :id
+
+  include Redis::Objects
+  EXPIRATION_INTERVAL = REDIS_CONFIG.fetch('client_session_timeout', 300)
+  hash_key :next_proxy_id, expiration: EXPIRATION_INTERVAL
+  hash_key :next_proxy_timestamp, expiration: EXPIRATION_INTERVAL
+
+  def initialize(id)
+    @id = id
+  end
+
+  def expiring_keys
+    [next_proxy_id, next_proxy_timestamp]
+  end
+
+  def valid?
+    expiring_keys.all?(&:exists?)
+  end
+
+  def touch
+    expiring_keys.each do |k|
+      k['keepalive'] = true
+      k.expire(EXPIRATION_INTERVAL)
+    end
+  end
+
+  def delete(site)
+    expiring_keys.each { |k| k.delete(site.id) }
+  end
+
+  def reserve_proxy(site, proxy_id, proxy_ts)
+    next_proxy_id[site.id] = proxy_id
+    next_proxy_timestamp[site.id] = proxy_ts
+    site.touch_proxy proxy_id
+    touch
+  end
+
+  def get_proxy(site, older_than)
+    proxy_ts = next_proxy_timestamp[site.id].andand.to_i
+    # We do not have a proxy reserved for this client/site combination.
+    # Find a proxy using the Site methods.
+    if proxy_ts.nil?
+      result = site.select_proxy(older_than)
+      # Cache a hot proxy for a later request.
+      if result.is_a? Proxy::NotReady
+        reserve_proxy site, result.proxy_id, result.proxy_ts
+      end
+    else
+      threshold_ts = Time.now.to_i - older_than
+      proxy_id = next_proxy_id[site.id]
+      # Our chached proxy is still too hot
+      if proxy_ts > threshold_ts
+        result = Proxy::NotReady.new(proxy_ts, threshold_ts, proxy_id)
+      # We have a chached proxy ready for use.
```

Comment:
(very) minor issue: should these comments be inside the blocks they refer to? It looks weird to me to interrupt an `if`/`else` with a comment.